taichi/transforms/auto_diff.cpp

```diff
@@ -181,20 +182,57 @@ class PromoteSSA2LocalVar : public BasicStmtVisitor {
     TI_ASSERT(stmt->width() == 1);
     if (!(stmt->is<UnaryOpStmt>() || stmt->is<BinaryOpStmt>() ||
           stmt->is<TernaryOpStmt>() || stmt->is<BitExtractStmt>() ||
-          stmt->is<GlobalLoadStmt>())) {
+          stmt->is<GlobalLoadStmt>() || stmt->is<AllocaStmt>())) {
       // TODO: this list may be incomplete
       return;
     }
-    // Create a alloc
-    auto alloc = Stmt::make<AllocaStmt>(1, stmt->ret_type);
-    auto alloc_ptr = alloc.get();
-    TI_ASSERT(alloca_block_);
-    alloca_block_->insert(std::move(alloc), 0);
-    auto load = stmt->insert_after_me(
-        Stmt::make<LocalLoadStmt>(LocalAddress(alloc_ptr, 0)));
-    irpass::replace_all_usages_with(stmt->parent, stmt, load);
-    // Create the load first so that the operand of the store won't get replaced
-    stmt->insert_after_me(Stmt::make<LocalStoreStmt>(alloc_ptr, stmt));
+
+    if (stmt->is<AllocaStmt>()) {
+      // Create a new alloc at the top of an ib to replace the old alloca
+      auto alloc = Stmt::make<AllocaStmt>(1, stmt->ret_type);
+      auto alloc_ptr = alloc.get();
+      TI_ASSERT(alloca_block_);
+      alloca_block_->insert(std::move(alloc), 0);
+      irpass::replace_all_usages_with(alloc_ptr->parent, alloc_ptr, stmt);
```

Comment:
OOC: why not just `irpass::replace_all_usages_with(stmt->parent, stmt, alloc_ptr);`? we can possible save the `gather_statements` below?